### ISSUE TITLE
Analyze support fixes and security fixes went for API-M 4.2.0, IS-6.0.0 and IS-6.1.0

### ISSUE THREAD

**Issue by HiranyaKavishani** [wso2/api-manager#2428]

**Comment by RusJaI** [comment]
Initial analysis was made checking for the PRs that has been made on each identity component. Following doc is containing the analysis results : 
https://docs.google.com/document/d/1NKZT4RaE25Te9YyDr_4chIw8GOn5OLhNErbmAdvjb84/edit?usp=sharing

After going through the corresponding patch tickets for each PR, it was identified that some fixes are applied across several repositories. (for example carbon-identity-inbound-oauth + carbon kernal).

With that, currently analysing the all set of PRs seperately along their patch entries to identify the other applicable PRs to the same fixes.

Other than that, there are Security patches which contains only manual file uploads which need to be analysed separately.

**Comment by RusJaI** [comment]
### Update [2024-01-29]

After an internal discussion on the scope of the fixes that need to be ported and their feasibility, it was decided to move to kernel version 4.9.x.

With that, as the initial effort, started checking the feasibility to build the dependent repos on kernel 4.9.x

**Comment by RusJaI** [comment]
### Update [2024-01-30]

There were build failures on some of the repositories due to the unavailability of the component correlation logging. 

In the progress of applying the corresponding component on carbon kernel 4.9.0 tag and building the kernel and the repos subsequently .

**Comment by RusJaI** [comment]
### Update  [2024-02-02]
Worked on the analysis of modified jsp files in **authenticationendpoint/includes**.
Listed the related PRs checking the diff between the files of U2 updated 4.2.0 live pack and Identity-apps 1.6.x branch.
Further checked the patches released modifying each of those jsp files in product IS and APIM 4.2.0.

Please find the sheet containing the analysis :https://docs.google.com/spreadsheets/d/1TN_q5sTZwm-Wb1S_dhKppJpTPszqittAhRW-gVAEGPc/edit?usp=sharing

**Comment by RusJaI** [comment]
### Update [2024-02-06, 07] - authenticationendpoint JSP file changes

Completed the analysis of IS related changes (from IS patches) for 1.6.x branch. Concluded that there are no identity fixes to be added to the 1.6.x branch.

Further identified several jsp files which are identical is IS and APIM, within authenticationendpoint/includes directory.

Built the APIM pack, excluding those fiels (this included building the carbon apimgt, identity oauth saml2, carbon kernel and product apim with some orbit jars which are not yet relleased.)

could build a product pack successfully, but the following error appears at the startup and the server stops. : 

```
org.wso2.andes.kernel.AndesException: Unable to initialise application registry
   at org.wso2.andes.server.Broker.startupImpl(Broker.java:312) ~[andes_3.3.28.jar:?]
   at org.wso2.andes.server.Broker.startup(Broker.java:110) ~[andes_3.3.28.jar:?]
   at org.wso2.andes.server.Main.startBroker(Main.java:218) ~[andes_3.3.28.jar:?]
   at org.wso2.andes.server.Main.execute(Main.java:207) ~[andes_3.3.28.jar:?]
   at org.wso2.andes.server.Main.<init>(Main.java:55) ~[andes_3.3.28.jar:?]
   at org.wso2.andes.server.Main.main(Main.java:48) ~[andes_3.3.28.jar:?]
   at org.wso2.carbon.andes.internal.QpidServiceComponent.startAndesBroker(QpidServiceComponent.java:351) ~[?:?]
   at org.wso2.carbon.andes.internal.QpidServiceComponent.activate(QpidServiceComponent.java:117) ~[?:?]
   at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
   at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
   at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke
```

Triggered another build upgrading the h2 orbit version in carbon apimgt and carbon kernel.

**Comment by RusJaI** [comment]
### Update [2024-02-08 ] - authenticationendpoint JSP file changes

Started working on merging the APIM customizations with updated jsp files in identity apps into product apim

Considered the diff which had been maintained in APIM 4.2.0 GA pack and the corresponding identity apps branch which was used there (1.5.0 tag). With that, merged the changes to product APIM, comparing it with updated identity apps branch (1.6.x)

completed the analysis for half of the files EoD

**Comment by RusJaI** [comment]
### Update [2024-02-09 ] - authenticationendpoint JSP file changes 

Identified another set of files other than the authenticationendpoint/includes, which needed to be synced with identity apps changes.
Completed the analysis and relevent Identity apps porting.

Worked on and Completed the process of merging identity apps changes aiming the alpha release.
(There are some patch changes to be ported, but cannot do at this moment because they have partial changes to other identity components which we haven't done porting up to now)

With that, jsp changes for alpha release is done. PR : https://github.com/wso2/identity-apps/pull/5461

Blocked in some build failures in product apim while trying to build a product pack with the changes (due to some other OSGI issues)
Hence tried applying the changed JSP files on an existing pack and tested the UI changes.

In the progress of removing backup-code usages from identity apps repo. The process seems not straightforward because there are errors from transitively affected files.

**Comment by RusJaI** [comment]
### Update [2024-02-10 ] - authenticationendpoint JSP file changes

Removed backup code usage with PR :  https://github.com/wso2/identity-apps/pull/5492
Now there are no initial startup errors looking for missing jsp files or features.
But when trying to navigate to login page of publisher portal, following error appears in the terminal and UI doesn't load properly.

```
[2024-02-12 21:08:55,349]  INFO - CarbonEventManagementService Starting polling event receivers
[2024-02-12 21:08:55,357] ERROR - [default] Servlet.service() for servlet [default] in context with path [/authenticationendpoint] threw exception
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 11 path $
	at com.google.gson.Gson.fromJson(Gson.java:1070) ~[com.google.gson_2.9.1.jar:?]
	at com.google.gson.Gson.fromJson(Gson.java:1016) ~[com.google.gson_2.9.1.jar:?]
	at com.google.gson.Gson.fromJson(Gson.java:959) ~[com.google.gson_2.9.1.jar:?]
	at com.google.gson.Gson.fromJson(Gson.java:927) ~[com.google.gson_2.9.1.jar:?]
	at org.wso2.carbon.identity.application.authentication.endpoint.util.filter.AuthParameterFilter.getServletRequestWithParams(AuthParameterFilter.java:117) ~[?:?]
	at org.wso2.carbon.identity.application.authentication.endpoint.util.filter.AuthParameterFilter.doFilter(AuthParameterFilter.java:82) ~[?:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:178) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:153) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.catalina.filters.HttpHeaderSecurityFilter.doFilter(HttpHeaderSecurityFilter.java:129) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:178) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:153) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:168) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:481) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:130) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.wso2.carbon.identity.context.rewrite.valve.TenantContextRewriteValve.invoke(TenantContextRewriteValve.java:119) ~[org.wso2.carbon.identity.context.rewrite.valve_1.8.11.jar:?]
	at org.wso2.carbon.identity.context.rewrite.valve.OrganizationContextRewriteValve.invoke(OrganizationContextRewriteValve.java:115) ~[org.wso2.carbon.identity.context.rewrite.valve_1.8.11.jar:?]
	at org.wso2.carbon.tomcat.ext.valves.SameSiteCookieValve.invoke(SameSiteCookieValve.java:38) ~[org.wso2.carbon.tomcat.ext_4.9.26.SNAPSHOT.jar:?]
	at org.wso2.carbon.identity.authz.valve.AuthorizationValve.invoke(AuthorizationValve.java:167) ~[org.wso2.carbon.identity.authz.valve_1.8.11.jar:?]
	at org.wso2.carbon.identity.auth.valve.AuthenticationValve.invoke(AuthenticationValve.java:118) ~[org.wso2.carbon.identity.auth.valve_1.8.11.jar:?]
	at org.wso2.carbon.tomcat.ext.valves.CompositeValve.continueInvocation(CompositeValve.java:114) ~[org.wso2.carbon.tomcat.ext_4.9.26.SNAPSHOT.jar:?]
	at org.wso2.carbon.tomcat.ext.valves.TomcatValveContainer.invokeValves(TomcatValveContainer.java:49) ~[org.wso2.carbon.tomcat.ext_4.9.26.SNAPSHOT.jar:?]
	at org.wso2.carbon.tomcat.ext.valves.CompositeValve.invoke(CompositeValve.java:75) ~[org.wso2.carbon.tomcat.ext_4.9.26.SNAPSHOT.jar:?]
	at org.wso2.carbon.tomcat.ext.valves.CarbonStuckThreadDetectionValve.invoke(CarbonStuckThreadDetectionValve.java:152) ~[org.wso2.carbon.tomcat.ext_4.9.26.SNAPSHOT.jar:?]
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:670) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.wso2.carbon.tomcat.ext.valves.CarbonContextCreatorValve.invoke(CarbonContextCreatorValve.java:63) ~[org.wso2.carbon.tomcat.ext_4.9.26.SNAPSHOT.jar:?]
	at org.wso2.carbon.tomcat.ext.valves.RequestCorrelationIdValve.invoke(RequestCorrelationIdValve.java:137) ~[org.wso2.carbon.tomcat.ext_4.9.26.SNAPSHOT.jar:?]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:390) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:928) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1794) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) ~[tomcat_9.0.82.wso2v1.jar:?]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) ~[tomcat_9.0.82.wso2v1.jar:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 11 path $
	at com.google.gson.stream.JsonReader.beginObject(JsonReader.java:395) ~[com.google.gson_2.9.1.jar:?]
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:182) ~[com.google.gson_2.9.1.jar:?]
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:144) ~[com.google.gson_2.9.1.jar:?]
	at com.google.gson.Gson.fromJson(Gson.java:1058) ~[com.google.gson_2.9.1.jar:?]
	... 38 more
```
When debugging the code, it was identified that, when it invokes the endpoint  authAPIURL =  https://localhost:9443/api/identity/auth/v1.1/data/AuthRequestKey/710dfce6-6460-449a-b2cd-5650ac0b1187, it returns the html page of the carbon console as contextProperties instead of the expected json response, at AuthParameterFilter.getServletRequestWithParams() in carbon-identity-framework.

**Comment by RusJaI** [comment]
### Update [2024-02-13, 14 ] - authenticationendpoint JSP file changes
When checking the identity framework tag in APIM 4.2.0, AuthParameterFilter.java class is not there. It has been introduced with a new IS feature.  Considering the said issue it was decided to remove the corresponding fix from 1.6.x branch. Created the PR : https://github.com/wso2/product-apim/pull/13371

Considering the impact of the said fix and after having another discussion with the IS team, it was decided to keep the fix in identity apps branch and merge it into product apim as well.

For M2 release it was targeted to keep the rest of the jsp changes + backup code removal.

Tried figuring out the build issue occured while building the product. Built all relevant repos in a clean repo and could resolve the build issue.

With that, built the APIM pack with jsp changes (not contain the security fix that caused the above issues), there was a problem in the loaded UI for login pages. fixed it by solving the name mismatch of a themes min css file.

When click on the forget password option it returned a 400 status code. When checked the same flow in M1 pack, the issue exists there as well. Created the issue https://github.com/wso2/api-manager/issues/2477 to track it.

With that, created the following PRs completing the targetted progress for alpha release : 

-  identity apps : https://github.com/wso2/identity-apps/pull/5492
- product apim : https://github.com/wso2/product-apim/pull/13371

**Comment by RusJaI** [comment]
Closing as completed
